py/private/py_pex_binary.bzl

```
@@ -0,0 +1,158 @@
+    # 
+    # No Match: `external/rules_python~~pip~pypi_39_rtoml/site-packages/rtoml-0.11.0/src/mod/parse.py`
+    # Reason: It has three `/` after first `site-packages` substring.
+    if site_packages_i != -1 and f.path.count("/", site_packages_i) == 2:
+        if f.path.find("dist-info", site_packages_i) != -1:
+            return ["--distinfo={}".format(f.dirname)]
+        return ["--dep={}".format(f.dirname)]
+
+    # If the path does not have a `site-packages` in it, then put it into 
+    # the standard runfiles tree.
+    elif site_packages_i == -1:
+        return ["--source={}={}".format(f.path, dest_path)]
+
+    return []
+
+def _py_python_pex_impl(ctx):
+    py_toolchain = _py_semantics.resolve_toolchain(ctx)
+
+    binary = ctx.attr.binary
+    runfiles = binary[DefaultInfo].data_runfiles
+
+    output = ctx.actions.declare_file(ctx.attr.name + ".pex")
+
+    args = ctx.actions.args()
+
+    # Copy workspace name here to prevent ctx
+    # being transferred to the execution phase.
+    workspace_name = str(ctx.workspace_name)
+
+    args.add_all(
+        ctx.attr.inject_env.items(), 
+        map_each = lambda e: "--inject-env=%s=%s" % (e[0], e[1]),
+        allow_closure = True,
+        # this is needed to allow passing a lambda (with workspace_name) to map_each 
+    )
+
+    args.add_all(
+        binary[PyInfo].imports, 
+        format_each = "--sys-path=%s"
+    )
+
+    args.add_all(
+        runfiles.files,
+        map_each = lambda f: _map_srcs(f, workspace_name),
+        uniquify = True,
+        # this is needed to allow passing a lambda (with workspace_name) to map_each 
+        allow_closure = True,
+    )
+    args.add(binary[DefaultInfo].files_to_run.executable, format = "--executable=%s")
+    args.add(ctx.attr.python_shebang, format = "--python-shebang=%s")
+    args.add(py_toolchain.python, format = "--python=%s")
+
+    py_version = py_toolchain.interpreter_version_info
+    args.add_all(
+        [
+            constraint.format(major = py_version.major, minor = py_version.minor, patch = py_version.micro) 
+            for constraint in ctx.attr.python_interpreter_constraints
+        ], 
+        format_each = "--python-version-constraint=%s"
+    )
+    args.add(output, format = "--output-file=%s")
+
+    ctx.actions.run(
+        executable = ctx.executable._pex,
+        inputs = runfiles.files,
+        arguments = [args],
+        outputs = [output],
+        mnemonic = "PyPex",
+        progress_message = "Building PEX binary %{label}",
+    )
+
+    return [
+        DefaultInfo(files = depset([output]), executable = output),
+        # See: https://github.com/bazelbuild/bazel/blob/b4ab259fe1cba8a108f1dd30067ee357c7198509/src/main/starlark/builtins_bzl/common/python/py_executable_bazel.bzl#L265
+        OutputGroupInfo(
+            python_zip_file = depset([output])
+        )
+    ]
+
+
+_attrs = dict({
+    "binary": attr.label(executable = True, cfg = "target"),
```

```suggestion
    "binary": attr.label(executable = True, mandatory = True, cfg = "target"),
```
Building a pex without a `binary` doesn't make much sense.

---

I think that a PEX can be used as an environment only - IIUIR the entry point(s) are option in the PEX tool https://docs.pex-tool.org/buildingpex.html#specifying-entry-points

---

@njlr there is no `deps` attribute in this rule (and I don't think we need it), so setting `binary` is the only way to specify what goes into PEX. In any way, the rule currently fails if you don't specify `binary` attribute:
```
ERROR: [...]/cli_tool/BUILD.bazel:19:14: in py_pex_binary rule //cli_tool:empty_pex:
Traceback (most recent call last):
	File "[...]/external/aspect_rules_py~/py/private/py_pex_binary.bzl", line 61, column 22, in _py_python_pex_impl
		runfiles = binary[DefaultInfo].data_runfiles
Error: type 'NoneType' has no operator [](Provider)
```

---

done